hasktorch/src/Torch/Typed/Vision.hs

```diff
@@ -140,3 +162,372 @@ initMnist path = do
   testImagesBS <- decompressFile path "t10k-images-idx3-ubyte.gz"
   testLabelsBS <- decompressFile path "t10k-labels-idx1-ubyte.gz"
   return (MnistData imagesBS labelsBS, MnistData testImagesBS testLabelsBS)
+
+
+-- ImageFolder dataset 
+
+newtype ImageFolder (m :: Type -> Type) (device :: (D.DeviceType, Nat) ) (finalShape :: [Nat]) = ImageFolder { folderData :: FolderData }
+
+
+instance (KnownShape shape, All KnownNat shape, [n, 3, h, w] ~ shape, 1 <= n, KnownDevice device, Applicative m) => 
+  Dataset m (ImageFolder m device shape) Int (Tensor device 'D.Float shape, Tensor device 'D.Int64 '[n]) where
+  getItem ImageFolder{..} ix =  
+    let
+      [n, c, h, w] = shapeVal @shape
+      indexes = [ix * n .. (ix+1) * n - 1]
+      (imgs, labels) = case someShape indexes of
+              SomeShape (Proxy :: Proxy idxs) -> 
+                (getFolderImages @idxs folderData, 
+                 getFolderLabels @idxs folderData)
+    in pure (imgs, labels)
+
+  keys ImageFolder{..} = fromList [ 0 .. (sum . map Prelude.length . imageNames $ folderData) `Prelude.div` (natValI @n) - 1]
+
+getFolderImage :: 
+  forall shape dtype device w h.
+  (All KnownNat shape, KnownDevice device, KnownDType dtype,
+  '[1, 3, w, h] ~ shape) =>
+  FilePath ->
+  [FilePath] -> 
+  Int ->
+  IO (Tensor device dtype shape)
+getFolderImage path filePaths imgIdx = (readImageAsRGB8 @shape @dtype @device 
+                                      . (++) path
+                                      . (!!) filePaths
+                                      $ (imgIdx `Prelude.mod` Prelude.length filePaths)) >>= \case  
+                                          Left err -> throwIO . userError $ "Path contains non-image files"
+                                          Right tensor -> return tensor
+
+
+getFolderImages ::
+  forall idxs shape dtype device n w h.
+  (All KnownNat shape, KnownDType dtype, KnownDevice device,
+   KnownShape idxs,
+   1 <= n,
+  '[n, 3, w, h] ~ shape) =>
+  FolderData ->
+  Tensor device dtype shape
+getFolderImages folder = squeezeDim @0
+                            . stack' @1 @'[1, 3, w, h] @(1 ': shape)
```

Comment:
ok, I see now why you want `stack'`. Alternative suggestion:
use `vecStack . fromJust . Data.Vector.Sized.fromList` here. at least it's explicit this way that this can crash.